Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/StartRide/Internal.hs

```diff
@@ -14,41 +14,69 @@
 
+import qualified Data.Time.Clock.POSIX as Time
 import qualified Domain.Types.Booking as SRB
 import qualified Domain.Types.Driver.DriverFlowStatus as DDFS
+import Domain.Types.Location as DLocation
 import qualified Domain.Types.Merchant as Dmerch
 import qualified Domain.Types.Person as SP
 import qualified Domain.Types.Ride as SRide
 import EulerHS.Prelude hiding (id)
-import Kernel.External.Maps.Types (LatLong)
+import Kernel.External.Maps
 import qualified Kernel.Storage.Esqueleto as Esq
 import Kernel.Storage.Esqueleto.Config (EsqLocDBFlow)
 import Kernel.Storage.Esqueleto.Transactionable (runInLocationDB)
+import Kernel.Storage.Hedis as Hedis
 import Kernel.Types.Common
 import Kernel.Types.Id
 import Lib.SessionizerMetrics.Types.Event
+import qualified SharedLogic.GoogleMaps as GoogleMaps
 import qualified SharedLogic.Ride as CQRide
 import Storage.CachedQueries.CacheConfig (CacheFlow)
 import qualified Storage.Queries.BusinessEvent as QBE
 import qualified Storage.Queries.Driver.DriverFlowStatus as QDFS
 import qualified Storage.Queries.DriverLocation as DrLoc
+import qualified Storage.Queries.Location as QLocation
+import qualified Storage.Queries.LocationMapping as QLocationMapping
 import qualified Storage.Queries.Ride as QRide
 import Tools.Event
+import qualified Tools.Maps as Maps
+import qualified Tools.Metrics as Metrics
 
 startRideTransaction ::
-  ( CacheFlow m r,
+  ( Metrics.CoreMetrics m,
+    CacheFlow m r,
+    EncFlow m r,
+    Hedis.HedisFlow m r,
     EsqDBFlow m r,
     EsqLocDBFlow m r,
+    Esq.EsqDBReplicaFlow m r,
+    HasField "minTripDistanceForReferralCfg" r (Maybe HighPrecMeters),
+    HasField "maxShards" r Int,
     EventStreamFlow m r
   ) =>
   Id SP.Person ->
   SRide.Ride ->
   Id SRB.Booking ->
   LatLong ->
   Id Dmerch.Merchant ->
-  m ()
+  m (Maybe DLocation.Location)
 startRideTransaction driverId ride bookingId firstPoint merchantId = do
   triggerRideStartEvent RideEventData {ride = ride{status = SRide.INPROGRESS}, personId = driverId, merchantId = merchantId}
+  startLocationCustomer <- do
+    fromLocationResponse <- Maps.getTripPlaceName merchantId Maps.GetPlaceNameReq {getBy = ByLatLong LatLong {lat = firstPoint.lat, lon = firstPoint.lon}, sessionToken = Nothing, language = Nothing}
+    actualFromLocationId <- generateGUID
+    fromAddress <- GoogleMaps.mkLocation fromLocationResponse
+    actualFromLocation <- DLocation.buildLocation actualFromLocationId firstPoint fromAddress
+    mappers <- QLocationMapping.findByTagId $ ride.id.getId
+    let toLocationMappers = filter (\mapper -> mapper.order /= 0) mappers
+    Esq.runTransaction $ QLocation.create actualFromLocation
+    epochVersion <- liftIO $ round `fmap` Time.getPOSIXTime
+    let epochVersionLast5Digits = epochVersion `mod` 100000 :: Integer
+    for_ toLocationMappers $ \locMap -> do
+      Esq.runTransaction $ QLocationMapping.updateLocationInMapping locMap actualFromLocation epochVersionLast5Digits
```

Comment:
this version calculation duplicated a lot. We can refactor, and move it inside of query function:

```
module Storage.Queries.LocationMapping

updateLocationInMapping :: Domain.LocationMapping -> Id Domain.Location -> SqlDB ()
updateLocationInMapping mapping newLocationId = do
  version <- Domain.getMappingVersion
  Esq.update $ \tbl -> do
    set
      tbl
      [ LocationMappingLocationId =. val (toKey newLocationId),
        LocationMappingVersion =. val (show version)
      ]
    where_ $ tbl ^. LocationMappingTId ==. val (toKey mapping.id)

module Domain.Types.LocationMapping

getMappingVersion :: MonadTime m => m Text
getMappingVersion = do
  now <- getCurrentTime
  let epochVersion = round $ Time.utcTimeToPOSIXSeconds now
  let epochVersionLast5Digits = epochVersion `mod` 100000 :: Integer
  pure $ show epochVersionLast5Digits
```